### ISSUE TITLE
Failure cluster [1ee05d1b...] - Requesting container logs with --tail=1 sometimes returns 2 rows

### ISSUE THREAD

**Issue by brianpursley** [kubernetes/kubernetes#95089]
##### Error text:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:698
Sep 18 18:02:34.531: Expected
    <int>: 2
to equal
    <int>: 1
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:1339
```
#### Recent failures:
[9/20/2020, 6:54:21 PM ci-kubernetes-e2e-ubuntu-gce](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-ubuntu-gce/1307815322513838080)
[9/20/2020, 12:09:20 PM ci-containerd-e2e-gci-gce](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-containerd-e2e-gci-gce/1307713401186684928)
[9/19/2020, 7:22:24 PM ci-cri-containerd-e2e-gce-multizone](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-cri-containerd-e2e-gce-multizone/1307459983889141760)
[9/19/2020, 5:15:22 PM ci-kubernetes-e2e-gci-gce-network-proxy-http-connect](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-network-proxy-http-connect/1307428022294417408)
[9/19/2020, 2:12:22 PM ci-kubernetes-e2e-gci-gce](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1307381969977348096)

/kind failing-test
/kind flake

Triage dashboard thinks this might be sig-cli, but I think it actually falls under another sig, I'm guessing sig-node.
/sig node

#### What I think the problem is

When a container is writing a high volume of log messages, there is a race condition between the time when the [start index is determined](https://github.com/kubernetes/kubernetes/blob/b5a02c419041f49200281778ae8f3d98a0d3df75/pkg/kubelet/kuberuntime/logs/logs.go#L298) and [the log file is read](https://github.com/kubernetes/kubernetes/blob/b5a02c419041f49200281778ae8f3d98a0d3df75/pkg/kubelet/kuberuntime/logs/logs.go#L320), and extra lines arrive in the log before the log file during that time, resulting in extra lines in the output sometimes.

#### How to reproduce the problem
Here is a script I wrote to reproduce the problem. It creates a pod that logs a lot of rows quickly, and then calls `kubectl log` using `--tail=2`, expecting 2 rows in the output.  My script uses `--tail=2` and sometimes gets 3 lines, but the concept is the same as the flaking test above, where it is using `--tail=1` and sometimes getting 2 lines.
```
kubectl apply -f <(cat << EOF
apiVersion: v1
kind: Pod
metadata:
  name: fastlogger 
spec:
  containers:
  - name: fastlogger
    image: busybox
    command: ["/bin/sh", "-c"]
    args: ["while true; do date +%s%N; done"]
EOF
)

kubectl wait --for=condition=Ready pod/fastlogger

for i in {1..10}
do
  linecount=$(kubectl logs fastlogger fastlogger --tail=2 --timestamps | wc -l)
  echo "linecount = $linecount"
  if [ $linecount -ne 2 ]; then
    echo "*** WRONG LINE COUNT ***"
  fi
done

kubectl delete pod fastlogger
```

Here is an example of the output produced, showing sometimes 3 rows are returned...
```
pod/fastlogger created
pod/fastlogger condition met
linecount = 2
linecount = 2
linecount = 3
*** WRONG LINE COUNT ***
linecount = 2
linecount = 3
*** WRONG LINE COUNT ***
linecount = 2
linecount = 2
linecount = 3
*** WRONG LINE COUNT ***
linecount = 2
linecount = 2
pod "fastlogger" deleted
```

**Comment by zhouya0** [comment]
I had a PR https://github.com/kubernetes/kubernetes/pull/93920 before. But stuck with unit test. If someone has figured out a better way. I'll closed my PR.

**Comment by fromanirh** [comment]
I see two relevant PRs (https://github.com/kubernetes/kubernetes/pull/95099 and https://github.com/kubernetes/kubernetes/pull/93920) both merged. Anything else needed to close this issue?

**Comment by brianpursley** [comment]
Hi @fromanirh 

It looks like @zhouya0 fixed the issue with https://github.com/kubernetes/kubernetes/pull/93920 which I think is in 1.21.

I tried to reproduce this issue on a 1.21 cluster and no longer see the problem occur.

I think this can be closed.

**Comment by fromanirh** [comment]
> Hi @fromanirh
> 
> It looks like @zhouya0 fixed the issue with #93920 which I think is in 1.21.
> 
> I tried to reproduce this issue on a 1.21 cluster and no longer see the problem occur.
> 
> I think this can be closed.

Thanks! :)
